resolwe/flow/views/annotations.py

```diff
@@ -0,0 +1,263 @@
+    AnnotationValueFilter,
+)
+from resolwe.flow.models import AnnotationPreset, Collection
+from resolwe.flow.models.annotations import (
+    AnnotationField,
+    AnnotationGroup,
+    AnnotationValue,
+)
+from resolwe.flow.serializers import (
+    AnnotationFieldSerializer,
+    AnnotationGroupSerializer,
+    AnnotationPresetSerializer,
+)
+from resolwe.flow.serializers.annotations import AnnotationValueSerializer
+from resolwe.permissions.loader import get_permissions_class
+from resolwe.permissions.mixins import ResolwePermissionsMixin
+from resolwe.permissions.models import Permission
+
+from .mixins import ResolweCreateModelMixin, ResolweUpdateModelMixin
+
+
+class AnnotationPresetViewSet(
+    ResolweCreateModelMixin,
+    mixins.RetrieveModelMixin,
+    ResolweUpdateModelMixin,
+    mixins.ListModelMixin,
+    ResolwePermissionsMixin,
+    viewsets.GenericViewSet,
+):
+    """API view for :class:`DescriptorSchema` objects."""
+
+    serializer_class = AnnotationPresetSerializer
+    permission_classes = (get_permissions_class(),)
+    filterset_class = AnnotationPresetFilter
+
+    ordering_fields = ("id", "name", "version")
+    ordering = ["sort_order"]
+    queryset = AnnotationPreset.objects.all()
+
+    def get_queryset(self):
+        """Get the presets for the current user."""
+        return super().get_queryset().filter_for_user(self.request.user)
+
+
+class AnnotationGroupViewSet(
+    mixins.RetrieveModelMixin,
+    mixins.ListModelMixin,
+    viewsets.GenericViewSet,
+):
+    """API view for AnnotationGroup objects."""
+
+    permission_classes = (permissions.IsAuthenticated,)
+    serializer_class = AnnotationGroupSerializer
+    filterset_class = AnnotationGroupFilter
+    ordering_fields = ("id", "name")
+    ordering = ["sort_order"]
+    queryset = AnnotationGroup.objects.all()
+
+
+class AnnotationFieldViewSet(
+    mixins.RetrieveModelMixin,
+    mixins.ListModelMixin,
+    viewsets.GenericViewSet,
+):
+    """Annotation fields viewset."""
+
+    permission_classes = (permissions.IsAuthenticated,)
+    serializer_class = AnnotationFieldSerializer
+    filterset_class = AnnotationFieldFilter
+    ordering_fields = ("id", "name", "label")
+    ordering = ["sort_order"]
+    queryset = AnnotationField.objects.all()
+
+    def _get_collection(self, collection_id: int, permission: Permission) -> Collection:
+        """Get the collection.
+
+        :raises ValidationError: when the collection could not be retrieved.
+        """
+        collection: Collection = (
+            Collection.objects.filter(pk=collection_id)
+            .filter_for_user(self.request.user, Permission.EDIT)
+            .first()
+        )
+        if collection is None:
+            raise exceptions.ValidationError(
+                f"No permission '{permission}' on collection with id '{collection_id}'."
+            )
+        return collection
+
+    @action(detail=False, methods=["post"])
+    def add_to_collection(self, request):
```

Comment:
This doesn't look clean. Create a separate serializer for the validation. And use objects instead of just ids, as it will be much easier for clients working that way.

Reply:
I have added serializers to validate the data.